Delete removeId()

NyorJa opened the pull request:
## Summary of change

Remove the method removeId and then assign id null as default or we can remove it. 



## Related issue

Closes #196 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw)

amit1307 commented on src/main/java/com/karankumar/bookproject/backend/entity/BaseEntity.java:
```diff
@@ -30,14 +30,8 @@
     @GeneratedValue(strategy = GenerationType.SEQUENCE)
     @Getter
-    private Long id;
+    private Long id = null;
```
Do we need explicit null assignment?

NyorJa replied:
we can remove it

knjk04 replied:
@NyorJa, I noticed you marked this conversation as resolved. However, I think it would be better for a reviewer (Amit or myself) to do so when we can confirm that this has been changed (as per our contributing guidelines).

amit1307 commented on src/main/java/com/karankumar/bookproject/backend/service/AuthorService.java:
```diff
@@ -50,17 +51,19 @@ public Author findById(Long id) {
-            List<Long> matchingAuthors = findAll().stream()
+            List<Author> existingAuthors = findAll();
+            List<Long> matchingAuthors = existingAuthors.stream()
                 .map(BaseEntity::getId)
                 .filter(id -> id.equals(author.getId())).collect(Collectors.toList());
-            if (matchingAuthors.size() == 1) {
+            // Null checking will do since it will return 1 or null or empty result since ids are unique
+            if (!CollectionUtils.isEmpty(matchingAuthors)) {
                 // if an author with the same ID exists, set this incoming author ID to null so that
                 // a new row in the table is made rather than updating the row that has the same ID
                 LOGGER.log(Level.INFO, "Matching authorIds: " + matchingAuthors);
-                author.removeId();
+                authorRepository.save(new Author(author.getFirstName(), author.getLastName()));
```
I am not sure about this logic here, all we are trying to find out if the id of the author to be saved already exists in the DB? If that's the case, shouldn't we just call some method from repo like findById and pass the id of the author to be saved?

amit1307 replied:
Actually thinking a bit more about it, seems like in any case we always want to add a new row so shouldn't we just save it as a new author always without even checking anything?

NyorJa replied:
i can change the query to findById since we are just validating if the author existing. shall I proceed with the changes?

amit1307 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,50 @@
+        // then
+        LOGGER.log(Level.INFO, "savedAuthors: " + savedAuthors.size());
+
+        // it should save 4 authors with different ids
```
Don't think we need this comment

knjk04 replied:
This can be replaced with `// then`

amit1307 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Test
+    public void testSaveAndConfirmDuplicateNameWithDifferentId() {
```
Can we please add one more test as below:
1. Save an author
2. Get the saved author 
3. Save it again after getting from DB
4. Check there are 2 in DB

Difference in this scenario is we are trying to save an author that has id assigned and making sure its not updated.

NyorJa replied:
ok. will do some more unit test

knjk04 replied:
@NyorJa, could you comment the method to make it clear to someone else what is being tested (i.e. the last line in Amit's comment or something similar)?

amit1307 commented on src/main/java/com/karankumar/bookproject/backend/service/AuthorService.java:
```diff
@@ -48,20 +47,16 @@ public Author findById(Long id) {
+            Author existingAuthor = findById(author.getId());
+            if (existingAuthor != null) {
+                LOGGER.log(Level.INFO, "Matching authorIds: " + existingAuthor.getId());
+                author = new Author(author.getFirstName(), author.getLastName());
```
To me it looks much better than the previous implementation, @knjk04 what do you think?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+        savedAuthors = sut.findAll();
+
+        assertEquals(2, savedAuthors.size());
+        System.out.println(savedAuthors.size());
```
Could you please remove this line? We should know the size irrespective of whether this test passes (then 2) or fails (JUnit will tell us)?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+    private final AuthorRepository authorRepository;
+
+    @Autowired
+    public AuthorServiceTest(AuthorService sut, BookService bookService, AuthorRepository authorRepository) {
```
I don't think we need to autowire the `AuthorRepository`. We can call `deleteAll()` on the `AuthorService` instance

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+@IntegrationTest
+public class AuthorServiceTest {
+
+    private final AuthorService sut;
```
By `sut`, I know you mean 'system under test', but this may not be clear to others. Hence, could you please rename this to the more readable `authorService`?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+    @Test
+    public void testSaveAndConfirmDuplicateNameWithDifferentId() {
+
+        //given
```
I know this is minor, but could you add a space after the last forward slash?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+        Author author = new Author("Nyor", "Ja");
+        sut.save(author);
+
+        Author author1 = author;
```
Could you please rename this to `authorCopy`?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+        Author author1 = author;
+        sut.save(author1);
+
+        Author author2 = new Author("Papa", "Cologne");
+        sut.save(author2);
+        sut.save(author2);
```
I don't think we need this. Testing it with one author should suffice

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+
+
+        // it should save 4 authors with different ids
+        assertEquals(4, savedAuthors.size());
```
Instead of calling `findAll` and then `size()`, how about using the `count()` method in `AuthorService`?

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+        sut.save(author2);
+        sut.save(author2);
+
+        // when
```
In this `when` section, I think it would be better to have the `save()` methods here, so that we have:

- Given an author and another author with the same author
- When we try to save authors with the same name
- We get back [expected]

knjk04 commented on src/test/java/com/karankumar/bookproject/backend/service/AuthorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+
+        List<Author> savedAuthors = sut.findAll();
+
+        // im having a hard time using this, im experiencing LazyInitializationException
+//        Author existingAuthor = sut.findById(author.getId());
```
Try annotating the method with `@Transactional`

knjk04 replied:
Could you also add a `// when` comment above this line once you've fixed it?

NyorJa replied:
Hi @knjk04 ,

I try that also and still encountering an error

knjk04 replied:
Could you paste the code with the annotation here? It works for me

knjk04 reviewed (changes requested):
Thanks for working on this, @NyorJa!

I've left some suggestions for improvements

knjk04 commented:
@NyorJa I noticed a bug:

- Log in
- Go to 'My books'
- Pick a shelf
- Select a row
- Change the name of one of the authors
- Click 'save'
- See error


Stack trace: https://gist.github.com/knjk04/360e5f6acd35dedaa3c2c26b4d1bd2c3

`removeId()` was created to resolve this, so it would be good if this also fixed. Unfortunately, this error message is somewhat cryptic, so let us know if you need a hand with debugging this

NyorJa commented:
> @NyorJa I noticed a bug:
> 
> * Log in
> * Go to 'My books'
> * Pick a shelf
> * Select a row
> * Change the name of one of the authors
> * Click 'save'
> * See error
> 
> Stack trace: https://gist.github.com/knjk04/360e5f6acd35dedaa3c2c26b4d1bd2c3
> 
> `removeId()` was created to resolve this, so it would be good if this also fixed. Unfortunately, this error message is somewhat cryptic, so let us know if you need a hand with debugging this

Please provide the credentials so that i can access the ui part

knjk04 commented:
@NyorJa Please refer to the [setup](https://github.com/Project-Books/book-project#setup) section of the README. The credentials are there.

NyorJa commented:
Hi @knjk04 ,

I already encounter the error on my feature branch. I am not sure if this is related to backend since i do not see in the logs related to it. I encounter it also in the master branch which is updated from the parent repository

knjk04 commented:
@NyorJa You're right, ignore me :)

knjk04 commented:
@NyorJa Could you please resolve the merge conflict?

NyorJa commented:
> @NyorJa Thanks for resolving the merge conflict. Once you've addressed the feedback from the last code review, feel free to request another review

I already apply the review points. Please check again

knjk04 commented:
> I already apply the review points. Please check again

I think there's one more point pending. See https://github.com/Project-Books/book-project/pull/383/files#r499134918

knjk04 commented:
@NyorJa Could you please squash the third commit onwards into the second commit? Once this is done, we'll merge

NyorJa commented:
> @NyorJa Could you please squash the third commit onwards into the second commit? Once this is done, we'll merge

I already squash the commit @knjk04 . Please check

knjk04 commented:
@NyorJa 

![image](https://user-images.githubusercontent.com/11173328/95214970-5b452f80-07e8-11eb-8fc9-72592b93b31d.png)

So by squashing the third commit onwards into the second commit, you should have two commits. I can see that you currently have eight commits.

NyorJa commented:
U force push it..sorry its a wrong move

knjk04 commented:
While on you're feature branch

Step 1: `$ git rebase -i HEAD~8`

This will open your commits in interactive rebase mode in the default Git text editor. This may be Vi or Vim, so you may wish to change the default text editor beforehand if you prefer something else.

Step 2: Replace the words 'pick' with 'squash' on the lines that you want to squash with an earlier commit

Step 3: Save and quit

Step 4: Force push: `git push -f origin [branch-name]

NyorJa commented:
> While on you're feature branch
> 
> Step 1: `$ git rebase -i HEAD~8`
> 
> This will open your commits in interactive rebase mode in the default Git text editor. This may be Vi or Vim, so you may wish to change the default text editor beforehand if you prefer something else.
> 
> Step 2: Replace the words 'pick' with 'squash' on the lines that you want to squash with an earlier commit
> 
> Step 3: Save and quit
> 
> Step 4: Force push: `git push -f origin [branch-name]`
> 
> I'm running this at the top of my head, so I may have skipped a step. However, this is the general process

what i did is 
git rebase -i HEAD~2 and then force push it. sorry for the misunderstanding

knjk04 commented:
@NyorJa Are you happy to join our Slack workspace so that we can help you with this further?

NyorJa commented:
> @NyorJa Are you happy to join our Slack workspace so that we can help you with this further?

I think that would be much better

knjk04 commented:
@NyorJa It looks like you're now including lots of commits unrelated to this PR. Could you please undo this? As ever, if you need a hand with this, feel free to reach out to us on Slack

NyorJa commented:
> @NyorJa It looks like you're now including lots of commits unrelated to this PR. Could you please undo this? As ever, if you need a hand with this, feel free to reach out to us on Slack

Would it be possible if I submit a new PR? The feature branch of this PR is kinda broken so I think a new PR will resolve this

knjk04 commented:
@NyorJa It looks to me like you've fixed it. no?

NyorJa commented:
> @NyorJa It looks to me like you've fixed it. no?

i thinks so since what i did is a rebase and pull the changes from the parent repo.

knjk04 commented:
@NyorJa Thanks for squashing your commits & for your help on this PR!
